packages/autoloader/src/AutoloadGenerator.php

```diff
@@ -0,0 +1,348 @@
+				: md5( uniqid( '', true ) );
+		}
+		// Copy over the autoload.php file into autoload_packages.php.
+		file_put_contents( $vendorPath . '/autoload_packages.php', $this->getAutoloadPackageFile( $suffix ) );
+		$this->io->writeError( '<info>Generated autoload_packages.php</info>', true );
+
+	}
+
+	/**
+	 * Takes a classMap and returns the array string representation.
+	 *
+	 * @param array $classMap Map of all the package classes and paths and versions.
+	 *
+	 * @return string
+	 */
+	private function classMapToPHPArrayString( $classMap ) {
+		$classmapString = ' array( ';
+		ksort( $classMap );
+		foreach ( $classMap as $class => $code ) {
+			$classmapString .= '    ' . var_export( $class, true ) . ' => ' . $code;
+		}
+		$classmapString .= ");\n";
+		return $classmapString;
+	}
+
+	/**
+	 * This function differs from the composer parseAutoloadsType in that beside returning the path.
+	 * It also return the path and the version of a package.
+	 *
+	 * @param array            $packageMap Map of all the packages.
+	 * @param string           $type Type of autoloader to use, currently not used, since we only support psr-4.
+	 * @param PackageInterface $mainPackage Instance of the Package Object.
+	 *
+	 * @return array
+	 */
+	protected function parseAutoloadsType( array $packageMap, $type, PackageInterface $mainPackage ) {
+		$autoloads = array();
+		foreach ( $packageMap as $item ) {
+			list($package, $installPath) = $item;
+			$autoload                    = $package->getAutoload();
+
+			if ( $package === $mainPackage ) {
+				$autoload = array_merge_recursive( $autoload, $package->getDevAutoload() );
+			}
+
+			// Skip packages that are not 'psr-4' since we only support them for now.
+			if ( ! isset( $autoload['psr-4'] ) || ! is_array( $autoload['psr-4'] ) ) {
+				continue;
+			}
+
+			if ( null !== $package->getTargetDir() && $package !== $mainPackage ) {
+				$installPath = substr( $installPath, 0, -strlen( '/' . $package->getTargetDir() ) );
+			}
+			foreach ( $autoload['psr-4'] as $namespace => $paths ) {
+				foreach ( (array) $paths as $path ) {
+					$relativePath              = empty( $installPath ) ? ( empty( $path ) ? '.' : $path ) : $installPath . '/' . $path;
+					$autoloads[ $namespace ][] = array(
+						'path'    => $relativePath,
+						'version' => $package->getVersion(), // Version of the class comes from the package - should we try to parse it?
+					);
+				}
+			}
+		}
+		return $autoloads;
+	}
+
+	/**
+	 * Take the autoloads array and return the classMap that contains the path and the version for each namespace.
+	 *
+	 * @param array      $autoloads Array of autoload settings defined defined by the packages.
+	 * @param Filesystem $filesystem Filesystem class instance.
+	 * @param string     $vendorPath Path to the vendor directory.
+	 * @param string     $basePath Base Path.
+	 *
+	 * @return array $classMap
+	 */
+	private function getClassMapFromAutoloads( $autoloads, $filesystem, $vendorPath, $basePath ) {
```

Comment:
Should we type hint here?

```suggestion
	private function getClassMapFromAutoloads( array $autoloads, Filesystem $filesystem, string $vendorPath, string $basePath ) {
```